File: Mobzy-API/src/main/java/com/offz/spigot/mobzy/Pathfinders/Flying/PathfinderGoalIdleFlyAboveGround.java
package com.offz.spigot.mobzy.Pathfinders.Flying;

import com.offz.spigot.mobzy.Mobs.Types.FlyingMob;

import java.util.Random;

public class PathfinderGoalIdleFlyAboveGround extends PathfinderGoalIdleFly {
    private double maxHeight;
    private double radius;

    public PathfinderGoalIdleFlyAboveGround(FlyingMob flyingMob, int maxHeight, double radius) {
        super(flyingMob);
        this.maxHeight = maxHeight;
        this.radius = radius;
    }

    @Override
    public void c() {
        Random random = mob.getRandom();
        double dx = mob.locX + ((random.nextFloat() * 2 - 1) * radius);
        double dy = mob.locY + ((random.nextFloat()) * maxHeight); //make it more likely to fly down
        double dz = mob.locZ + ((random.nextFloat() * 2 - 1) * radius);

        if (mob.getBukkitEntity().getWorld().getBlockAt((int) dx, (int) (dy - maxHeight), (int) dz).getType().isSolid())
            mob.getControllerMove().a(dx, dy, dz, 1.0D);
        else
            mob.getControllerMove().a(dx, mob.locY - maxHeight, dz, 1.0D);
    }
}

File: Mobzy-API/src/main/java/com/offz/spigot/mobzy/pathfinders/flying/PathfinderGoalIdleFlyAboveGround.kt
package com.offz.spigot.mobzy.pathfinders.flying

import com.offz.spigot.mobzy.mobs.types.FlyingMob
import org.bukkit.Location
import kotlin.random.Random

class PathfinderGoalIdleFlyAboveGround(mob: FlyingMob, private val maxHeight: Double = 4.0, private val radius: Double = 5.0) : PathfinderGoalIdleFly(mob) {

    override fun init() {
        val x = mob.x + Random.nextDouble(-radius, radius)
        val y = mob.y + Random.nextDouble(-radius / 4, radius) //make it more likely to fly down
        val z = mob.z + Random.nextDouble(-radius, radius)
        val loc = Location(entity.world, x, y, z)
        if (!loc.clone().add(0.0, -maxHeight, 0.0).block.type.isSolid) {
            moveController.a(mob.x, mob.y - 0.1, mob.z, 1.0)
            return
        }
        if (y > 16) { //keep mobs from going down and killing themselves
            targetLoc = loc
            moveController.a(x, y, z, 1.0) //TODO make a wrapper for the controller and figure out the difference between it and navigation
        }
    }
}